In [2]:
import pandas as pd
import numpy as np

## Обзор данных

Для считывания данных используется функция `pd.read_csv`. У этой функции много аргументов (см. документацию), первый из которых - название считываемого файла с данными, второй - sep - разделитель, по умолчанию это запятая. Данные в csv файлах (обычно, но не всегда) записаны в виде "таблицы". Одна запись - одна строка, данные столбцов разделяются запятыми.

Для работы с файлом его, как и обычный текстовый файл (впрочем, это и есть обычный текстовый файл!), удобнее положить в ту же директорию, где лежит ноутбук. Затем вы можете прямо через jupyter открыть файл `csv` и увидеть, как он выглядит.

In [3]:
f = pd.read_csv("telecom_churn.csv")

Посмотреть начало и конец таблицы (по умолчанию 5 строк, но это настраиваемый параметр) - функции `head` и `tail`. Описание данных - `describe`, `info`. Размер таблицы - `shape`, названия колонок - `columns`. Обратите внимание, что `columns` возвращает итерируемый объект.

In [5]:
f.head()

State  Account length  Area code International plan Voice mail plan  \
0    KS             128        415                 No             Yes   
1    OH             107        415                 No             Yes   
2    NJ             137        415                 No              No   
3    OH              84        408                Yes              No   
4    OK              75        415                Yes              No   

   Number vmail messages  Total day minutes  Total day calls  \
0                     25              265.1              110   
1                     26              161.6              123   
2                      0              243.4              114   
3                      0              299.4               71   
4                      0              166.7              113   

   Total day charge  Total eve minutes  Total eve calls  Total eve charge  \
0             45.07              197.4               99             16.78   
1             27.47              195.5              103             16.62   
2             41.38              121.2              110             10.30   
3             50.90               61.9               88              5.26   
4             28.34              148.3              122             12.61   

   Total night minutes  Total night calls  Total night charge  \
0                244.7                 91               11.01   
1                254.4                103               11.45   
2                162.6                104                7.32   
3                196.9                 89                8.86   
4                186.9                121                8.41   

   Total intl minutes  Total intl calls  Total intl charge  \
0                10.0                 3               2.70   
1                13.7                 3               3.70   
2                12.2                 5               3.29   
3                 6.6                 7               1.78   
4                10.1                 3               2.73   

   Customer service calls  Churn  
0                       1  False  
1                       1  False  
2                       0  False  
3                       2  False  
4                       3  False

In [6]:
f.head(7)

State  Account length  Area code International plan Voice mail plan  \
0    KS             128        415                 No             Yes   
1    OH             107        415                 No             Yes   
2    NJ             137        415                 No              No   
3    OH              84        408                Yes              No   
4    OK              75        415                Yes              No   
5    AL             118        510                Yes              No   
6    MA             121        510                 No             Yes   

   Number vmail messages  Total day minutes  Total day calls  \
0                     25              265.1              110   
1                     26              161.6              123   
2                      0              243.4              114   
3                      0              299.4               71   
4                      0              166.7              113   
5                      0              223.4               98   
6                     24              218.2               88   

   Total day charge  Total eve minutes  Total eve calls  Total eve charge  \
0             45.07              197.4               99             16.78   
1             27.47              195.5              103             16.62   
2             41.38              121.2              110             10.30   
3             50.90               61.9               88              5.26   
4             28.34              148.3              122             12.61   
5             37.98              220.6              101             18.75   
6             37.09              348.5              108             29.62   

   Total night minutes  Total night calls  Total night charge  \
0                244.7                 91               11.01   
1                254.4                103               11.45   
2                162.6                104                7.32   
3                196.9                 89                8.86   
4                186.9                121                8.41   
5                203.9                118                9.18   
6                212.6                118                9.57   

   Total intl minutes  Total intl calls  Total intl charge  \
0                10.0                 3               2.70   
1                13.7                 3               3.70   
2                12.2                 5               3.29   
3                 6.6                 7               1.78   
4                10.1                 3               2.73   
5                 6.3                 6               1.70   
6                 7.5                 7               2.03   

   Customer service calls  Churn  
0                       1  False  
1                       1  False  
2                       0  False  
3                       2  False  
4                       3  False  
5                       0  False  
6                       3  False

In [ ]:
f.tail()

In [ ]:
f.tail(7)

In [ ]:
f.tail(100)

In [7]:
f.shape

(3333, 20)

In [ ]:
f.columns

In [8]:
f.describe()

Account length    Area code  Number vmail messages  Total day minutes  \
count     3333.000000  3333.000000            3333.000000        3333.000000   
mean       101.064806   437.182418               8.099010         179.775098   
std         39.822106    42.371290              13.688365          54.467389   
min          1.000000   408.000000               0.000000           0.000000   
25%         74.000000   408.000000               0.000000         143.700000   
50%        101.000000   415.000000               0.000000         179.400000   
75%        127.000000   510.000000              20.000000         216.400000   
max        243.000000   510.000000              51.000000         350.800000   

       Total day calls  Total day charge  Total eve minutes  Total eve calls  \
count      3333.000000       3333.000000        3333.000000      3333.000000   
mean        100.435644         30.562307         200.980348       100.114311   
std          20.069084          9.259435          50.713844        19.922625   
min           0.000000          0.000000           0.000000         0.000000   
25%          87.000000         24.430000         166.600000        87.000000   
50%         101.000000         30.500000         201.400000       100.000000   
75%         114.000000         36.790000         235.300000       114.000000   
max         165.000000         59.640000         363.700000       170.000000   

       Total eve charge  Total night minutes  Total night calls  \
count       3333.000000          3333.000000        3333.000000   
mean          17.083540           200.872037         100.107711   
std            4.310668            50.573847          19.568609   
min            0.000000            23.200000          33.000000   
25%           14.160000           167.000000          87.000000   
50%           17.120000           201.200000         100.000000   
75%           20.000000           235.300000         113.000000   
max           30.910000           395.000000         175.000000   

       Total night charge  Total intl minutes  Total intl calls  \
count         3333.000000         3333.000000       3333.000000   
mean             9.039325           10.237294          4.479448   
std              2.275873            2.791840          2.461214   
min              1.040000            0.000000          0.000000   
25%              7.520000            8.500000          3.000000   
50%              9.050000           10.300000          4.000000   
75%             10.590000           12.100000          6.000000   
max             17.770000           20.000000         20.000000   

       Total intl charge  Customer service calls  
count        3333.000000             3333.000000  
mean            2.764581                1.562856  
std             0.753773                1.315491  
min             0.000000                0.000000  
25%             2.300000                1.000000  
50%             2.780000                1.000000  
75%             3.270000                2.000000  
max             5.400000                9.000000

In [9]:
f.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3333 entries, 0 to 3332
Data columns (total 20 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   State                   3333 non-null   object 
 1   Account length          3333 non-null   int64  
 2   Area code               3333 non-null   int64  
 3   International plan      3333 non-null   object 
 4   Voice mail plan         3333 non-null   object 
 5   Number vmail messages   3333 non-null   int64  
 6   Total day minutes       3333 non-null   float64
 7   Total day calls         3333 non-null   int64  
 8   Total day charge        3333 non-null   float64
 9   Total eve minutes       3333 non-null   float64
 10  Total eve calls         3333 non-null   int64  
 11  Total eve charge        3333 non-null   float64
 12  Total night minutes     3333 non-null   float64
 13  Total night calls       3333 non-null   int64  
 14  Total night charge      3333 non-null   

Небольшое, но важное отступление. Глубоко понимать это пока не обязательно. Обратите внимание, что функция `print` возвращает строковое представление объекта и она не необходима, чтобы вывести объект. 2 ячейки сработают одинаково:

In [10]:
a = 1
print(a)

1


In [11]:
a = 1
a

1

При этом для некоторых объектов (для которых стандартные функции `__str__` и `__repr__` возвращают разные значения) два способа вывода сработают несколько по-разному:

In [12]:
a = np.array([1.0, 2.0])
print(a)

[1. 2.]


In [13]:
a = np.array([1.0, 2.0])
a

array([1., 2.])

Такая же ситуация с pandas dataframe'ами. Без print получится красивее, чем с ним, но фактически это одно и то же:

In [15]:
print(f)

     State  Account length  Area code International plan Voice mail plan  \
0       KS             128        415                 No             Yes   
1       OH             107        415                 No             Yes   
2       NJ             137        415                 No              No   
3       OH              84        408                Yes              No   
4       OK              75        415                Yes              No   
...    ...             ...        ...                ...             ...   
3328    AZ             192        415                 No             Yes   
3329    WV              68        415                 No              No   
3330    RI              28        510                 No              No   
3331    CT             184        510                Yes              No   
3332    TN              74        415                 No             Yes   

      Number vmail messages  Total day minutes  Total day calls  \
0                   

In [14]:
f

State  Account length  Area code International plan Voice mail plan  \
0       KS             128        415                 No             Yes   
1       OH             107        415                 No             Yes   
2       NJ             137        415                 No              No   
3       OH              84        408                Yes              No   
4       OK              75        415                Yes              No   
...    ...             ...        ...                ...             ...   
3328    AZ             192        415                 No             Yes   
3329    WV              68        415                 No              No   
3330    RI              28        510                 No              No   
3331    CT             184        510                Yes              No   
3332    TN              74        415                 No             Yes   

      Number vmail messages  Total day minutes  Total day calls  \
0                        25              265.1              110   
1                        26              161.6              123   
2                         0              243.4              114   
3                         0              299.4               71   
4                         0              166.7              113   
...                     ...                ...              ...   
3328                     36              156.2               77   
3329                      0              231.1               57   
3330                      0              180.8              109   
3331                      0              213.8              105   
3332                     25              234.4              113   

      Total day charge  Total eve minutes  Total eve calls  Total eve charge  \
0                45.07              197.4               99             16.78   
1                27.47              195.5              103             16.62   
2                41.38              121.2              110             10.30   
3                50.90               61.9               88              5.26   
4                28.34              148.3              122             12.61   
...                ...                ...              ...               ...   
3328             26.55              215.5              126             18.32   
3329             39.29              153.4               55             13.04   
3330             30.74              288.8               58             24.55   
3331             36.35              159.6               84             13.57   
3332             39.85              265.9               82             22.60   

      Total night minutes  Total night calls  Total night charge  \
0                   244.7                 91               11.01   
1                   254.4                103               11.45   
2                   162.6                104                7.32   
3                   196.9                 89                8.86   
4                   186.9                121                8.41   
...                   ...                ...                 ...   
3328                279.1                 83               12.56   
3329                191.3                123                8.61   
3330                191.9                 91                8.64   
3331                139.2                137                6.26   
3332                241.4                 77               10.86   

      Total intl minutes  Total intl calls  Total intl charge  \
0                   10.0                 3               2.70   
1                   13.7                 3               3.70   
2                   12.2                 5               3.29   
3                    6.6                 7               1.78   
4                   10.1                 3               2.73   
...                  ...               ...                ...   
3328                 9.9                 6            

2 варианта синтаксиса для просмотра содержимого столбца:

In [16]:
f.State

0       KS
1       OH
2       NJ
3       OH
4       OK
        ..
3328    AZ
3329    WV
3330    RI
3331    CT
3332    TN
Name: State, Length: 3333, dtype: object

In [17]:
f['State']

0       KS
1       OH
2       NJ
3       OH
4       OK
        ..
3328    AZ
3329    WV
3330    RI
3331    CT
3332    TN
Name: State, Length: 3333, dtype: object

Обратите внимание, что таблица и столбцы имеют разный тип:

In [18]:
type(f), type(f.State), type(f['State'])

(pandas.core.frame.DataFrame,
 pandas.core.series.Series,
 pandas.core.series.Series)

Тем не менее, мы можем сделать именно датафрейм из одного столбца следующим образом:

In [19]:
f[['State']]

State
0       KS
1       OH
2       NJ
3       OH
4       OK
...    ...
3328    AZ
3329    WV
3330    RI
3331    CT
3332    TN

[3333 rows x 1 columns]

In [ ]:
type(f[['State']]), type(f['State'])

Это способ обращения к "срезам" таблицы по названиям столбцов. Еще пример:

In [ ]:
f[['State', 'Churn']]

Просмотр уникальных значений в столбце и их числа:

In [ ]:
f.State.unique()

In [ ]:
f.State.nunique()

Обращение к элементу ообъекта `pd.Series`:

In [ ]:
f.State[0]

Полный набор функций, применимых к столбцу можно посмотреть следующим образом. Здесь используется стандартная функция `__dir__()`, которая выводит список атрибутов и методов объекта.

In [ ]:
f.State.__dir__()

С `pd.DataFrame` привычный способ обращения по индексам не сработает:

In [ ]:
f[0][0]

Основной способ - использование `loc` и `iloc`. `loc` дает индексацию по "настоящим" индексам - названиям столбцов и названиям строк. В нашем случае, названия строк - целые числа, как обычные индексы. `iloc` позволяет осуществлять стандартную численную индексацию, как для `numpy` массивов.

In [ ]:
f.iloc[0] # первая строка

In [ ]:
f.iloc[0][0]

In [ ]:
f.iloc[0, 0]

In [ ]:
f.iloc[[0, 0]]

In [ ]:
f.iloc[[1, 2, 5, 1]]

In [ ]:
f.loc[[1, 2, 5, 1]]

In [ ]:
f.loc[[1, 2, 5, 1], ["State", "Account length"]]

In [ ]:
f.iloc[[1, 2, 5, 1], [0, 1]]

Следующая ячейка вернет ошибку:

In [ ]:
f.iloc[[1, 2, 5, 1], ["State", "Account length"]]

In [ ]:
f.iloc[:5, :4]

In [ ]:
f.loc[:, ["State", "Account length"]]

In [ ]:
f.loc[4:9, ["State", "Account length", "Area code", "International plan"]]

Обратите внимание, что `loc` включает и строки соответствующие и начальному, и конечному индексам.

Сохраним этот датафрейм в новую переменную и рассмотрим отличие loc и iloc в работе по целочисленным индексам строк

In [21]:
new_f = f.loc[4:9, ["State", "Account length", "Area code", "International plan"]]

In [22]:
new_f.iloc[:5]

State  Account length  Area code International plan
4    OK              75        415                Yes
5    AL             118        510                Yes
6    MA             121        510                 No
7    MO             147        415                Yes
8    LA             117        408                 No

In [23]:
new_f.loc[:5]

State  Account length  Area code International plan
4    OK              75        415                Yes
5    AL             118        510                Yes

**Упражнение 1.** Объясните различие поведения `loc` и `iloc` для `new_f`.

#### Создание датафрейма и присваивание по индексам

In [ ]:
df = pd.DataFrame({"A": [1, 2, 3], "B": [4, 5, 6]}, index = ["first", "second", "third"])
df

In [ ]:
df.iloc[0, 0] = 100
df

In [ ]:
df.loc["second", "B"] = 31
df

In [ ]:
df.iloc[2] = 5
df

In [ ]:
df.loc["B"] = 0
df

In [ ]:
df.loc[:, "B"] = 0
df

In [ ]:
df = pd.DataFrame({"A": [1, 2, 3, 4], "B": [5, 6, 7, 8]})
df

In [ ]:
df.iloc[1:2, :] = -1
df

In [ ]:
df.loc[1:2, :] = -2
df

Еще раз убеждаемся, что `loc` включает оба конца "среза", а `iloc` - полуинтервал, как например обычные срезы.

#### Создание нового столбца

In [ ]:
df["C"] = 1
df

In [ ]:
df["D"] = df["A"] + df["B"]
df

In [ ]:
df["E"] = df["D"].apply(lambda x: x ** 2)
df

In [ ]:
df["F"] = df[["A", "C", "E"]].apply(lambda x: x.sum(), axis=1)
df

**Упражнение 2.** Объясните последний результат.

## Срезы с условиями и аггрегация

In [32]:
f[f.Churn == True]

State  Account length  Area code International plan Voice mail plan  \
10      IN              65        415                 No              No   
15      NY             161        415                 No              No   
21      CO              77        408                 No              No   
33      AZ              12        408                 No              No   
41      MD             135        408                Yes             Yes   
...    ...             ...        ...                ...             ...   
3301    CA              84        415                 No              No   
3304    IL              71        510                Yes              No   
3320    GA             122        510                Yes              No   
3322    MD              62        408                 No              No   
3323    IN             117        415                 No              No   

      Number vmail messages  Total day minutes  Total day calls  \
10                        0              129.1              137   
15                        0              332.9               67   
21                        0               62.4               89   
33                        0              249.6              118   
41                       41              173.1               85   
...                     ...                ...              ...   
3301                      0              280.0              113   
3304                      0              186.1              114   
3320                      0              140.0              101   
3322                      0              321.1              105   
3323                      0              118.4              126   

      Total day charge  Total eve minutes  Total eve calls  Total eve charge  \
10               21.95              228.5               83             19.42   
15               56.59              317.8               97             27.01   
21               10.61              169.9              121             14.44   
33               42.43              252.4              119             21.45   
41               29.43              203.9              107             17.33   
...                ...                ...              ...               ...   
3301             47.60              202.2               90             17.19   
3304             31.64              198.6              140             16.88   
3320             23.80              196.4               77             16.69   
3322             54.59              265.5              122             22.57   
3323             20.13              249.3               97             21.19   

      Total night minutes  Total night calls  Total night charge  \
10                  208.8                111                9.40   
15                  160.6                128                7.23   
21                  209.6                 64                9.43   
33                  280.2                 90               12.61   
41                  122.2                 78                5.50   
...                   ...                ...                 ...   
3301                156.8                103                7.06   
3304                206.5                 80                9.29   
3320                120.1                133                5.40   
3322                180.5                 72                8.12   
3323                227.0                 56               10.22   

      Total intl minutes  Total intl calls  Total intl charge  \
10                  12.7                 6               3.43   
15                   5.4                 9               1.46   
21                   5.7                 6               1.54   
33                  11.8                 3               3.19   
41                  14.6                15               3.94   
...                  ...               ...                ...   
3301                10.4                 4            

In [33]:
f[f["Area code"] == 408]

State  Account length  Area code International plan Voice mail plan  \
3       OH              84        408                Yes              No   
8       LA             117        408                 No              No   
12      IA             168        408                 No              No   
16      ID              85        408                 No             Yes   
21      CO              77        408                 No              No   
...    ...             ...        ...                ...             ...   
3309    VT             100        408                Yes              No   
3312    SC             181        408                 No              No   
3313    ID             127        408                 No              No   
3322    MD              62        408                 No              No   
3325    OH              78        408                 No              No   

      Number vmail messages  Total day minutes  Total day calls  \
3                         0              299.4               71   
8                         0              184.5               97   
12                        0              128.8               96   
16                       27              196.4              139   
21                        0               62.4               89   
...                     ...                ...              ...   
3309                      0              219.4              112   
3312                      0              229.9              130   
3313                      0              102.8              128   
3322                      0              321.1              105   
3325                      0              193.4               99   

      Total day charge  Total eve minutes  Total eve calls  Total eve charge  \
3                50.90               61.9               88              5.26   
8                31.37              351.6               80             29.89   
12               21.90              104.9               71              8.92   
16               33.39              280.9               90             23.88   
21               10.61              169.9              121             14.44   
...                ...                ...              ...               ...   
3309             37.30              225.7              102             19.18   
3312             39.08              144.4               93             12.27   
3313             17.48              143.7               95             12.21   
3322             54.59              265.5              122             22.57   
3325             32.88              116.9               88              9.94   

      Total night minutes  Total night calls  Total night charge  \
3                   196.9                 89                8.86   
8                   215.8                 90                9.71   
12                  141.1                128                6.35   
16                   89.3                 75                4.02   
21                  209.6                 64                9.43   
...                   ...                ...                 ...   
3309                255.3                 95               11.49   
3312                262.4                110               11.81   
3313                191.4                 97                8.61   
3322                180.5                 72                8.12   
3325                243.3                109               10.95   

      Total intl minutes  Total intl calls  Total intl charge  \
3                    6.6                 7               1.78   
8                    8.7                 4               2.35   
12                  11.2                 2               3.02   
16                  13.8                 4               3.73   
21                   5.7                 6               1.54   
...                  ...               ...                ...   
3309                12.0                 4            

In [34]:
f[f["Area code"] == 408].loc[3]

State                        OH
Account length               84
Area code                   408
International plan          Yes
Voice mail plan              No
Number vmail messages         0
Total day minutes         299.4
Total day calls              71
Total day charge           50.9
Total eve minutes          61.9
Total eve calls              88
Total eve charge           5.26
Total night minutes       196.9
Total night calls            89
Total night charge         8.86
Total intl minutes          6.6
Total intl calls              7
Total intl charge          1.78
Customer service calls        2
Churn                     False
Name: 3, dtype: object

In [35]:
f[f["Area code"] == 408].loc[3] == f[f["Area code"] == 408].iloc[0]

State                     True
Account length            True
Area code                 True
International plan        True
Voice mail plan           True
Number vmail messages     True
Total day minutes         True
Total day calls           True
Total day charge          True
Total eve minutes         True
Total eve calls           True
Total eve charge          True
Total night minutes       True
Total night calls         True
Total night charge        True
Total intl minutes        True
Total intl calls          True
Total intl charge         True
Customer service calls    True
Churn                     True
Name: 3, dtype: bool

In [16]:
f[f.Churn == True]

State  Account length  Area code International plan Voice mail plan  \
10      IN              65        415                 No              No   
15      NY             161        415                 No              No   
21      CO              77        408                 No              No   
33      AZ              12        408                 No              No   
41      MD             135        408                Yes             Yes   
...    ...             ...        ...                ...             ...   
3301    CA              84        415                 No              No   
3304    IL              71        510                Yes              No   
3320    GA             122        510                Yes              No   
3322    MD              62        408                 No              No   
3323    IN             117        415                 No              No   

      Number vmail messages  Total day minutes  Total day calls  \
10                        0              129.1              137   
15                        0              332.9               67   
21                        0               62.4               89   
33                        0              249.6              118   
41                       41              173.1               85   
...                     ...                ...              ...   
3301                      0              280.0              113   
3304                      0              186.1              114   
3320                      0              140.0              101   
3322                      0              321.1              105   
3323                      0              118.4              126   

      Total day charge  Total eve minutes  Total eve calls  Total eve charge  \
10               21.95              228.5               83             19.42   
15               56.59              317.8               97             27.01   
21               10.61              169.9              121             14.44   
33               42.43              252.4              119             21.45   
41               29.43              203.9              107             17.33   
...                ...                ...              ...               ...   
3301             47.60              202.2               90             17.19   
3304             31.64              198.6              140             16.88   
3320             23.80              196.4               77             16.69   
3322             54.59              265.5              122             22.57   
3323             20.13              249.3               97             21.19   

      Total night minutes  Total night calls  Total night charge  \
10                  208.8                111                9.40   
15                  160.6                128                7.23   
21                  209.6                 64                9.43   
33                  280.2                 90               12.61   
41                  122.2                 78                5.50   
...                   ...                ...                 ...   
3301                156.8                103                7.06   
3304                206.5                 80                9.29   
3320                120.1                133                5.40   
3322                180.5                 72                8.12   
3323                227.0                 56               10.22   

      Total intl minutes  Total intl calls  Total intl charge  \
10                  12.7                 6               3.43   
15                   5.4                 9               1.46   
21                   5.7                 6               1.54   
33                  11.8                 3               3.19   
41                  14.6                15               3.94   
...                  ...               ...                ...   
3301                10.4                 4            

In [ ]:
f[f.Churn == True].mean()

In [ ]:
f[f.Churn == True].sum()

Выборка по нескольким условиям требует скобок, т.к. операции `&`, `|`, `~` (и, или, не) имеют более высокий приоритет, чем `==`, `>=`, `<=` и арифметические:

In [ ]:
f[(f.Churn == True) & (f["Area code"] == 408)]

Можем посмотреть средние показатели лояльных и нелояльных клиентов (столбец "Churn" показывает, ушел клиент от нас как от оператора связи или нет).

In [ ]:
loyal = f[f.Churn == False].mean()
unloyal = f[f.Churn == True].mean()
pd.DataFrame({"loyal": loyal, "unloyal": unloyal})

`groupby` позволяет делать группировку по уникальным значениям столбца, которые будут играть роль индекса в полученном датафрейме. При этом также нужна функция агрегации (`mean`, `sum`, `max` и прочие), чтобы отобразить полученный объект.

In [ ]:
f.groupby("Churn")

In [ ]:
f.groupby("Churn").mean()

Здесь мы сделали агрегацию всех столбцов функцией `mean`. Часто нам нужна агрегация по конкретному или конкретным столбцам. Посмотрим, например, в каком штате больше всего разговаривают по телефону:

In [ ]:
f.groupby(["State"]).agg({"Total day minutes": "mean"})

Максимальное значение времени:

In [ ]:
f.groupby(["State"]).agg({"Total day minutes": "mean"}).max()

Чтобы найти сам штат, можем отсортировать данные:

In [ ]:
f.groupby(["State"]).agg({"Total day minutes": "mean"}).sort_values(by="Total day minutes", ascending = False)

**Пример.** Как распределен отток клиентов по штатам? В каких штатах он больше среднего?

In [42]:
f.groupby(["State"]).agg({"Churn": "mean"})

In [43]:
f.groupby(["State"]).agg({"Churn": "mean"}).loc["LA"]

Churn    0.078431
Name: LA, dtype: float64

In [ ]:
f.groupby(["State"]).agg({"Churn": "mean"}).mean()

In [ ]:
f.groupby(["State"]).agg({"Churn": "mean"}).mean()[0]

In [4]:
m = f.groupby(["State"]).agg({"Churn": "mean"}).mean()[0]
new_dataframe = f.groupby(["State"]).agg({"Churn": "mean"})
new_dataframe[new_dataframe.Churn > m]

Churn
State          
AR     0.200000
CA     0.264706
CT     0.162162
DE     0.147541
GA     0.148148
KS     0.185714
MA     0.169231
MD     0.242857
ME     0.209677
MI     0.219178
MN     0.178571
MS     0.215385
MT     0.205882
NC     0.161765
NH     0.160714
NJ     0.264706
NV     0.212121
NY     0.180723
OK     0.147541
PA     0.177778
SC     0.233333
TX     0.250000
WA     0.212121

## Упражнения.

**Упражнение 3.** Найдите среднее количество звонков `Total day calls` для всего датафрейма.

In [29]:
print(np.mean(f.loc[:, 'Total day calls']))

100.43564356435644


**Упражнение 4.** Найдите среднее количество звонков `Total day calls` для любого выбранного вами штата.

In [45]:
state = "OH"
print(f.groupby(["State"]).agg({"Total day calls": "mean"}).loc[state])

Total day calls    99.628205
Name: OH, dtype: float64


**Упражнение 5.** Создайте датафрейм, в котором будет среднее количество звонков `Total day calls` для каждого штата.

In [6]:
day_calls_f = f.groupby(["State"]).agg({"Total day calls": "mean"})
day_calls_f

Total day calls
State                 
AK           95.269231
AL           99.100000
AR          102.145455
AZ          100.109375
CA          104.500000
CO           95.121212
CT           98.432432
DC          100.166667
DE           98.639344
FL          101.000000
GA          103.055556
HI          102.811321
IA          103.045455
ID          106.219178
IL          101.689655
IN          107.056338
KS          101.028571
KY          100.338983
LA          103.509804
MA          102.292308
MD           99.357143
ME           99.500000
MI          101.205479
MN           99.119048
MO           99.936508
MS          100.707692
MT           95.220588
NC          102.794118
ND          100.145161
NE           97.409836
NH           98.964286
NJ          100.102941
NM          102.241935
NV          101.424242
NY           98.240964
OH           99.628205
OK          100.262295
OR           95.730769
PA           98.644444
RI          100.430769
SC          102.466667
SD           94.816667
TN          101.452830
TX           99.986111
UT           98.958333
VA          105.363636
VT          100.328767
WA          101.196970
WI           96.679487
WV          103.783019
WY          102.220779

**Упражнение 6.** Оставьте в созданном датафрейме строки только с теми штатами, где количество звонков `Total day calls` больше среднего по исходному датафрейму.

In [19]:
day_calls_mean = np.mean(f.loc[:, "Total day calls"])
day_calls_f = day_calls_f.loc[day_calls_f["Total day calls"] > day_calls_mean, :]
day_calls_f

Total day calls
State                 
AR          102.145455
CA          104.500000
FL          101.000000
GA          103.055556
HI          102.811321
IA          103.045455
ID          106.219178
IL          101.689655
IN          107.056338
KS          101.028571
LA          103.509804
MA          102.292308
MI          101.205479
MS          100.707692
NC          102.794118
NM          102.241935
NV          101.424242
SC          102.466667
TN          101.452830
VA          105.363636
WA          101.196970
WV          103.783019
WY          102.220779

**Упражнение 7.** Создайте датафрейм, в котором будует средние количества звонков `Total day calls` и `Total eve calls` для каждого штата.

In [20]:
day_eve_calls_f = f.groupby(["State"]).agg({"Total day calls": "mean",
                                            "Total eve calls": "mean"})
day_eve_calls_f

Total day calls  Total eve calls
State                                  
AK           95.269231        96.576923
AL           99.100000        98.462500
AR          102.145455        97.836364
AZ          100.109375       102.406250
CA          104.500000       103.088235
CO           95.121212        98.166667
CT           98.432432        99.121622
DC          100.166667       101.981481
DE           98.639344        97.360656
FL          101.000000       108.920635
GA          103.055556       103.888889
HI          102.811321        98.849057
IA          103.045455        99.545455
ID          106.219178       100.054795
IL          101.689655       100.810345
IN          107.056338        97.943662
KS          101.028571       101.528571
KY          100.338983        98.237288
LA          103.509804        96.627451
MA          102.292308        98.230769
MD           99.357143        98.571429
ME           99.500000       104.596774
MI          101.205479       101.986301
MN           99.119048       101.642857
MO           99.936508       101.047619
MS          100.707692       100.200000
MT           95.220588        97.514706
NC          102.794118       103.941176
ND          100.145161       105.467742
NE           97.409836       103.590164
NH           98.964286        95.803571
NJ          100.102941       103.441176
NM          102.241935       102.983871
NV          101.424242        96.257576
NY           98.240964       100.024096
OH           99.628205       101.205128
OK          100.262295       104.278689
OR           95.730769        98.500000
PA           98.644444       101.622222
RI          100.430769        94.969231
SC          102.466667        99.783333
SD           94.816667        98.633333
TN          101.452830       103.754717
TX           99.986111       101.902778
UT           98.958333        98.027778
VA          105.363636        99.948052
VT          100.328767        97.671233
WA          101.196970        95.742424
WI           96.679487        99.538462
WV          103.783019        98.056604
WY          102.220779        99.233766

**Упражнение 8.** Создайте датафрейм, в котором будует средние количества звонков `Total day calls` и `Total eve calls` для каждого штата, а также столбец со значениями True и False - ответом на вопрос, больше ли дневных звонков, чем вечерних.

In [23]:
day_eve_calls_f.loc[:, "day > eve"] = day_eve_calls_f.loc[:, "Total day calls"]\
    > day_eve_calls_f.loc[:, "Total eve calls"]
day_eve_calls_f

Total day calls  Total eve calls  day > eve
State                                             
AK           95.269231        96.576923      False
AL           99.100000        98.462500       True
AR          102.145455        97.836364       True
AZ          100.109375       102.406250      False
CA          104.500000       103.088235       True
CO           95.121212        98.166667      False
CT           98.432432        99.121622      False
DC          100.166667       101.981481      False
DE           98.639344        97.360656       True
FL          101.000000       108.920635      False
GA          103.055556       103.888889      False
HI          102.811321        98.849057       True
IA          103.045455        99.545455       True
ID          106.219178       100.054795       True
IL          101.689655       100.810345       True
IN          107.056338        97.943662       True
KS          101.028571       101.528571      False
KY          100.338983        98.237288       True
LA          103.509804        96.627451       True
MA          102.292308        98.230769       True
MD           99.357143        98.571429       True
ME           99.500000       104.596774      False
MI          101.205479       101.986301      False
MN           99.119048       101.642857      False
MO           99.936508       101.047619      False
MS          100.707692       100.200000       True
MT           95.220588        97.514706      False
NC          102.794118       103.941176      False
ND          100.145161       105.467742      False
NE           97.409836       103.590164      False
NH           98.964286        95.803571       True
NJ          100.102941       103.441176      False
NM          102.241935       102.983871      False
NV          101.424242        96.257576       True
NY           98.240964       100.024096      False
OH           99.628205       101.205128      False
OK          100.262295       104.278689      False
OR           95.730769        98.500000      False
PA           98.644444       101.622222      False
RI          100.430769        94.969231       True
SC          102.466667        99.783333       True
SD           94.816667        98.633333      False
TN          101.452830       103.754717      False
TX           99.986111       101.902778      False
UT           98.958333        98.027778       True
VA          105.363636        99.948052       True
VT          100.328767        97.671233       True
WA          101.196970        95.742424       True
WI           96.679487        99.538462      False
WV          103.783019        98.056604       True
WY          102.220779        99.233766       True

**Упражнение 9.** Найти долю клиентов (отношение их числа к общему количеству клиентов) с `international plan` и `voice mail plan`.

In [34]:
ip_vmp_ratio =\
    len(f.loc[f.loc[:, "International plan"] == "Yes", "International plan"]) /\
    len(f.loc[f.loc[:, "Voice mail plan"] == "Yes", "Voice mail plan"])
print(ip_vmp_ratio)

0.35032537960954446


**Упражнение 10.** Найти число уникальных значений `Area code`.

In [35]:
unique_area_code = f.loc[:, "Area code"].unique()
print(unique_area_code)

[415 408 510]


**Упражнение 11.** Вывести DataFrame из 2 столбцов: число звонков в поддержку; число клиентов, звонивших столько раз. Подсказка: используйте функцию агрегации `count`.

State  Account length  Area code International plan Voice mail plan  \
0       KS             128        415                 No             Yes   
1       OH             107        415                 No             Yes   
2       NJ             137        415                 No              No   
3       OH              84        408                Yes              No   
4       OK              75        415                Yes              No   
...    ...             ...        ...                ...             ...   
3328    AZ             192        415                 No             Yes   
3329    WV              68        415                 No              No   
3330    RI              28        510                 No              No   
3331    CT             184        510                Yes              No   
3332    TN              74        415                 No             Yes   

      Number vmail messages  Total day minutes  Total day calls  \
0                        25              265.1              110   
1                        26              161.6              123   
2                         0              243.4              114   
3                         0              299.4               71   
4                         0              166.7              113   
...                     ...                ...              ...   
3328                     36              156.2               77   
3329                      0              231.1               57   
3330                      0              180.8              109   
3331                      0              213.8              105   
3332                     25              234.4              113   

      Total day charge  Total eve minutes  Total eve calls  Total eve charge  \
0                45.07              197.4               99             16.78   
1                27.47              195.5              103             16.62   
2                41.38              121.2              110             10.30   
3                50.90               61.9               88              5.26   
4                28.34              148.3              122             12.61   
...                ...                ...              ...               ...   
3328             26.55              215.5              126             18.32   
3329             39.29              153.4               55             13.04   
3330             30.74              288.8               58             24.55   
3331             36.35              159.6               84             13.57   
3332             39.85              265.9               82             22.60   

      Total night minutes  Total night calls  Total night charge  \
0                   244.7                 91               11.01   
1                   254.4                103               11.45   
2                   162.6                104                7.32   
3                   196.9                 89                8.86   
4                   186.9                121                8.41   
...                   ...                ...                 ...   
3328                279.1                 83               12.56   
3329                191.3                123                8.61   
3330                191.9                 91                8.64   
3331                139.2                137                6.26   
3332                241.4                 77               10.86   

      Total intl minutes  Total intl calls  Total intl charge  \
0                   10.0                 3               2.70   
1                   13.7                 3               3.70   
2                   12.2                 5               3.29   
3                    6.6                 7               1.78   
4                   10.1                 3               2.73   
...                  ...               ...                ...   
3328                 9.9                 6            

**Упражнение 12.** Вывести DataFrame из 2 столбцов: число звонков в поддержку; доля оттока (`Churn`). Построить график.

**Упражнение 13.** Найти среднюю длительность международного (`intl`) звонка.

**Упражнение 14*.** Какие звонки дольше - дневные, вечерние или ночные? Ответ привести в формате DataFrame $3*3$: строки - число минут, число звонков, среднее время звонка. Столбцы - день, вечер, ночь

**Упражнение 15.** Сравнить `Total day charge` для оставшихся и ушедших клиентов.

**Упражнение 16.** Отсортриуйте штаты по `Total day charge` (по возрастанию).

**Упражнение 17.** Сделайте агрегацию по средним показателям для каждой `Area code`.

**Упражнение 18.** Выведите датафрейм размера $3*2$: столбцы State, Churn для 100, 102 и 104 строк исходного датафрейма.

**Упражнение 19.** Создайте датафрейм из 2 столбцов, заполненных произвольными числами. Добавьте третий столбец, равный их сумме квадратов.

**Упражнение 20.** Добавьте к созданному датафрейму четвертый столбец, равный среднему значению первых трех, *используя* функцию `mean`.